src/modules/m_clones.cpp

```diff
@@ -23,16 +23,97 @@
 
 /** Handle /CLONES
  */
+
+
+typedef TR1NS::unordered_set<User*> UList;
+typedef std::map<irc::sockets::cidr_mask, UList > CloneList;
+
+class UserTracker : public Timer
+{
+	bool tracking;
+	unsigned long timeout;
+	CloneList cl;
+
+public:
+	UserTracker()
+		: Timer(timeout, false), tracking(false), timeout(0)
+	{
+	}
+
+	bool Tick(time_t) CXX11_OVERRIDE
+	{
+		cl.clear();
+		tracking = false;
+		return false;
+	}
+
+	void SetTimeout(unsigned long t)
+	{
+		timeout = t;
+	}
+
+	CloneList& GetMap()
+	{
+		if (tracking)
+		{
+			SetInterval(timeout);
+			return cl;
+		}
+
+		if (timeout)
+		{
+			tracking = true;
+			SetInterval(timeout);
+		}
+
+		for (user_hash::const_iterator it = ServerInstance->Users->clientlist.begin(); it != ServerInstance->Users->clientlist.end(); ++it)
+		{
+			User* user = &(*it->second);
+			AddClient(user);
+		}
+		return cl;
+	}
+
+	void AddClient(User* user)
+	{
+		if (!tracking)
+			return;
+
+		UList& counts = cl[user->GetCIDRMask()];
+		counts.insert(user);
+	}
+
+	void DelClient(User* user)
+	{
+		if (!tracking)
+			return;
+
+		CloneList::iterator it = cl.find(user->GetCIDRMask());
+		if (it != cl.end())
+		{
+			UList& counts = it->second;
+			counts.erase(user);
+			// If there's no more users from this IP, remove entry from the map
+			if (counts.empty())
+				cl.erase(it);
+		}
+	}
+};
+
 class CommandClones : public Command
 {
+	UserTracker& utrack;
+
  public:
- 	CommandClones(Module* Creator) : Command(Creator,"CLONES", 1)
+	CommandClones(Module* Creator, UserTracker& ut)
+		: Command(Creator,"CLONES", 1), utrack(ut)
 	{
 		flags_needed = 'o'; syntax = "<limit>";
 	}
 
 	CmdResult Handle (const std::vector<std::string> &parameters, User *user)
 	{
+		CloneList& clonemap = utrack.GetMap();
```

Comment:
Move this near first use (that is, before the loop like the current code does with the clonemap)